### PR TITLE
http: Implement various missing commands

### PR DESCRIPTION
**Pull request by DaniElectra**
This PR implements various HTTP commands that some applications use to connect to their servers. Specifically:

- `GetRequestState`
- `AddPostDataBianry`
- `SetPostDataType`
- All `SendPostData` commands
- `SetPostDataEncoding`
- `GetResponseHeaderTimeout`
- `SetPostDataTypeSize`

This allows games that don't require Nintendo servers to connect online, like Flipnote Studio 3D with the Kaeru World servers patch. It also prepares ground work for those that do, as some games will perform HTTP requests to get and post user generated content.

To be honest I haven't looked much into reverse engineered code of the sysmodule, but I've made lots of tests with various edge cases on the original sysmodule and tried to replicate those as accurately as possible.

Screenshot:
imagen

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7415)

### CODE REVIEW

File: src/core/hle/service/http/http_c.cpp [hunk 1]
```diff
@@ -130,8 +173,38 @@ static std::size_t WriteHeaders(httplib::Stream& stream,
...
-                                     httplib::Stream& strm, httplib::Headers& httplib_headers) {
+static void SerializeChunkedAsciiPostData(httplib::DataSink& sink, Context::Params& params) {
+    std::string query;
+
+    for (auto it = params.begin(); it != params.end(); ++it) {
+        if (it != params.begin()) {
+            sink.os << "&";
+        }
+
+        query = it->first;
```
Comment (SachinVin) [inline src/core/hle/service/http/http_c.cpp:184]: I think it would be better to use fmt::format

File: src/core/hle/service/http/http_c.cpp [hunk 2]
```diff
@@ -34,36 +35,78 @@ enum {
...
     SessionStateError = 102,
+
+    WrongCertHandle = 201,
     TooManyClientCerts = 203,
     NotImplemented = 1012,
 };
 }
 
-const Result ERROR_STATE_ERROR = // 0xD8A0A066
+constexpr Result ErrorStateError = // 0xD8A0A066
```
[the remark this answers is not in the document]
Comment (DaniElectra) [reply src/core/hle/service/http/http_c.cpp:62]: @GPUCode has previously told to use PascalCase for results: https://github.com/citra-emu/citra/pull/7377#discussion_r1461169300

File: src/core/hle/service/http/http_c.cpp [hunk 3]
```diff
@@ -774,21 +1061,359 @@ void HTTP_C::AddPostDataRaw(Kernel::HLERequestContext& ctx) {
...
+    Kernel::MappedBuffer& value_buffer = rp.PopMappedBuffer();
+
+    // Copy the name_buffer into a string without the \0 at the end
+    const std::string name(name_buffer.begin(), name_buffer.end() - 1);
+
+    // Copy the value_buffer into a string without the \0 at the end
+    std::string value(value_size - 1, '\0');
+    value_buffer.Read(value.data(), 0, value_size - 1);
+
+    if (timeout) {
```
[the remark this answers is not in the document]
Comment (DaniElectra) [reply src/core/hle/service/http/http_c.cpp:1138]: These are two different commands, but using the same function as it mostly behaves the same (hence the `Impl` function name). The `timeout` argument allows distinguishing them

File: src/core/hle/service/http/http_c.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SachinVin) [reply src/core/hle/service/http/http_c.cpp:62]: I see, i think it would be good to have a comment about it here

[the remark this answers is not in the document]
Comment (GPUCode) [reply src/core/hle/service/http/http_c.cpp:62]: The entire codebase has been switched to PascalCase results so if you mean a comment about the case I don't know if it's necessary

File: src/core/hle/service/http/http_c.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (SachinVin) [reply src/core/hle/service/http/http_c.cpp:1138]: Ah, i probably would have used a single log with `timeout` logged, but this is ok

File: src/core/hle/service/http/http_c.cpp [hunk 4]
```diff
@@ -130,8 +173,38 @@ static std::size_t WriteHeaders(httplib::Stream& stream,
     return write_len;
 }
 
-static std::size_t HandleHeaderWrite(std::vector<Context::RequestHeader>& pending_headers,
-                                     httplib::Stream& strm, httplib::Headers& httplib_headers) {
+static void SerializeChunkedAsciiPostData(httplib::DataSink& sink, Context::Params& params) {
+    std::string query;
+
+    for (auto it = params.begin(); it != params.end(); ++it) {
```
[the remark this answers is not in the document]
Comment (DaniElectra) [reply src/core/hle/service/http/http_c.cpp:179]: I have to check below if the element isn't the first element to add the `&` chunk. Is it possible to do this with a for each loop?

File: src/core/hle/service/http/http_c.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (GPUCode) [reply src/core/hle/service/http/http_c.cpp:62]: This sounds like something suited for the code style guidelines page

File: src/core/hle/service/http/http_c.cpp (hunk 1 above)
Comment (DaniElectra) [reply src/core/hle/service/http/http_c.cpp:184]: Do you mean this? `query = fmt::format("{}", it->first);`

Comment (DaniElectra) [reply src/core/hle/service/http/http_c.cpp:184]: oh i see now, disregard

File: src/core/hle/service/http/http_c.cpp [hunk 5]
```diff
@@ -234,16 +377,15 @@ void Context::MakeRequestNonSSL(httplib::Request& request, const URLInfo& url_in
         std::make_unique<httplib::Client>(url_info.host, url_info.port);
 
     client->set_header_writer(
-        [&pending_headers](httplib::Stream& strm, httplib::Headers& httplib_headers) {
-            return HandleHeaderWrite(pending_headers, strm, httplib_headers);
+        [this, &pending_headers](httplib::Stream& strm, httplib::Headers& httplib_headers) {
+            return this->HandleHeaderWrite(pending_headers, strm, httplib_headers);
```
Comment (Steveice10) [inline src/core/hle/service/http/http_c.cpp:381]: Is the `this->` here required? I thought as long as `this` is captured it works fine without.

File: src/core/hle/service/http/http_c.cpp [hunk 6]
```diff
@@ -293,20 +435,58 @@ void Context::MakeRequestSSL(httplib::Request& request, const URLInfo& url_info,
     client->enable_server_certificate_verification(false);
 
     client->set_header_writer(
-        [&pending_headers](httplib::Stream& strm, httplib::Headers& httplib_headers) {
-            return HandleHeaderWrite(pending_headers, strm, httplib_headers);
+        [this, &pending_headers](httplib::Stream& strm, httplib::Headers& httplib_headers) {
+            return this->HandleHeaderWrite(pending_headers, strm, httplib_headers);
```
Comment (Steveice10) [inline src/core/hle/service/http/http_c.cpp:439]: Ditto here regarding `this->`.

**Review by Steveice10 — COMMENTED** [review]
Looks good, thanks for this. Just a small nit from me.

File: src/core/hle/service/http/http_c.h [hunk 7]
```diff
@@ -197,6 +211,41 @@ class Context final {
         friend class boost::serialization::access;
     };
 
+    struct Param {
+        Param(std::vector<u8> value)
```
[the remark this answers is not in the document]
Comment (SachinVin) [reply src/core/hle/service/http/http_c.h]: Const reference as in `Param(const std::vector<u8>& value)`

File: src/core/hle/service/http/http_c.cpp [hunk 8]
```diff
@@ -130,8 +173,38 @@ static std::size_t WriteHeaders(httplib::Stream& stream,
...
+            fmt::format("{}={}", it->first, httplib::detail::encode_query_param(it->second.value));
+        boost::replace_all(query, "*", "%2A");
+        sink.os << query;
+    }
+}
+
+static void SerializeChunkedMultipartPostData(httplib::DataSink& sink,
+                                              const Context::Params& params,
+                                              const std::string& boundary) {
+    std::string body;
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:194]: the variable declaration can be moved into the loop it seems

File: src/core/hle/service/http/http_c.cpp [hunk 9]
```diff
@@ -130,8 +173,38 @@ static std::size_t WriteHeaders(httplib::Stream& stream,
...
+
+    for (const auto& param : params) {
+        httplib::MultipartFormData item = param.second.ToMultipartForm();
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:197]:
```suggestion
        const auto item = param.second.ToMultipartForm();
```

File: src/core/hle/service/http/http_c.cpp [hunk 10]
```diff
@@ -293,20 +435,58 @@ void Context::MakeRequestSSL(httplib::Request& request, const URLInfo& url_info,
...
         LOG_ERROR(Service_HTTP, "Request failed: {}: {}", error, httplib::to_string(error));
         state = RequestState::TimedOut;
     } else {
         LOG_DEBUG(Service_HTTP, "Request successful");
-        // TODO(B3N30): Verify this state on HW
         state = RequestState::ReadyToDownloadContent;
     }
 }
 
+bool Context::ContentProvider(size_t offset, size_t length, httplib::DataSink& sink) {
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:451]: the function name could be a little more descriptive. Is this setting the ContentProvider?
Also it seems like it always returns `true`.

Same with the function below,

File: src/core/hle/service/http/http_c.cpp [hunk 11]
```diff
@@ -774,21 +1061,359 @@ void HTTP_C::AddPostDataRaw(Kernel::HLERequestContext& ctx) {
...
+    SendPostDataAsciiImpl(ctx, true);
+}
+
+void HTTP_C::SendPostDataAsciiImpl(Kernel::HLERequestContext& ctx, bool timeout) {
+    IPC::RequestParser rp(ctx);
+    const u32 context_handle = rp.Pop<u32>();
+    [[maybe_unused]] const u32 name_size = rp.Pop<u32>();
+    const u32 value_size = rp.Pop<u32>();
+    // TODO(DaniElectra): The original module waits until a connection with the server is made
+    u64 timeout_nanos;
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:1124]:
```suggestion
    const u64 timeout_nanos = timeout ? rp.Pop<u64>() : 0;
```

File: src/core/hle/service/http/http_c.cpp [hunk 12]
```diff
@@ -774,21 +1061,359 @@ void HTTP_C::AddPostDataRaw(Kernel::HLERequestContext& ctx) {
...
+    SendPostDataBinaryImpl(ctx, true);
+}
+
+void HTTP_C::SendPostDataBinaryImpl(Kernel::HLERequestContext& ctx, bool timeout) {
+    IPC::RequestParser rp(ctx);
+    const u32 context_handle = rp.Pop<u32>();
+    [[maybe_unused]] const u32 name_size = rp.Pop<u32>();
+    const u32 value_size = rp.Pop<u32>();
+    // TODO(DaniElectra): The original module waits until a connection with the server is made
+    u64 timeout_nanos;
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:1191]:
```suggestion
     const u64 timeout_nanos = timeout ? rp.Pop<u64>() : 0;
```

File: src/core/hle/service/http/http_c.cpp [hunk 13]
```diff
@@ -774,21 +1061,359 @@ void HTTP_C::AddPostDataRaw(Kernel::HLERequestContext& ctx) {
...
+void HTTP_C::SendPostDataRawTimeout(Kernel::HLERequestContext& ctx) {
+    SendPostDataRawImpl(ctx, true);
+}
+
+void HTTP_C::SendPostDataRawImpl(Kernel::HLERequestContext& ctx, bool timeout) {
+    IPC::RequestParser rp(ctx);
+    const u32 context_handle = rp.Pop<u32>();
+    const u32 post_data_len = rp.Pop<u32>();
+    // TODO(DaniElectra): The original module waits until a connection with the server is made
+    u64 timeout_nanos;
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:1257]:
```suggestion
     const u64 timeout_nanos = timeout ? rp.Pop<u64>() : 0;
```

File: src/core/hle/service/http/http_c.cpp [hunk 14]
```diff
@@ -774,21 +1061,359 @@ void HTTP_C::AddPostDataRaw(Kernel::HLERequestContext& ctx) {
...
+    if (timeout) {
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.cpp:1258]: now that im thinking about it, you could always log the timeout bool variable and then get rid of the two different log statements in the functions

File: src/core/hle/service/http/http_c.h [hunk 15]
```diff
@@ -197,6 +211,41 @@ class Context final {
         friend class boost::serialization::access;
     };
 
+    struct Param {
+        Param(const std::vector<u8> value)
+            : name(value.begin(), value.end()), value(value.begin(), value.end()){};
```
Comment (FearlessTobi) [inline src/core/hle/service/http/http_c.h:216]: in other places you made sure to exclude the null-terminator when constructing a string. Is this not required here?

File: src/core/hle/service/http/http_c.cpp (hunk 10 above)
Comment (DaniElectra) [reply src/core/hle/service/http/http_c.cpp:451]: This function is the "content provider" used by httplib to get the HTTP request body, we also use it to update the request state accordingly when the function is called. [httplib](https://github.com/yhirose/cpp-httplib?tab=readme-ov-file#send-content-with-a-content-provider) refers to it by that name, so that's what I used.

We have two versions: one for normal requests and another for *chunked* requests `ChunkedContentProvider`

> Also it seems like it always returns true.

As per the httplib documentation, the function returns a bool that tells wether to cancel the request `false` or to continue `true`. We don't have a use case where we would need to cancel a request in the middle of the HTTP request AFAIK, so we always return true.

File: src/core/hle/service/http/http_c.h (hunk 15 above)
Comment (DaniElectra) [reply src/core/hle/service/http/http_c.h:216]: The `std::vector<u8>` value here won't be a string, but a raw or binary value (these are different use cases: raw is for the raw HTTP body, binary is a binary value for a multipart form).

In this specific function, this is a workaround to store the `SendPostDataRaw(Timeout)` value as a `Param` to support a specific quirk of the HTTP sysmodule: if you use `PostDataType::Raw`, you can concatenate all three types of `SendPostData` values and those will be added to the final body of the request. I can add an `is_binary(true)` here for clarity, even though it won't be used.

**Review by Steveice10 — APPROVED** [review]
